core/src/upickle/core/Util.scala

```diff
@@ -72,25 +72,26 @@ object Util {
     var inverseSign: Long = -1L
     var i: Int = start
 
+    if ((start | end | end - start | (cs.length - end)) < 0) throw new IndexOutOfBoundsException
```

Comment:
Sorry, I can't understand this check here.
You make the or of the `start`, `end` etc. and check if the result is `< 0`. Why?
Also are you sure there isn't redundancy between this check and  the one in line 83?

Reply:
I think I got the meaning. You do the OR of all these numbers since none of them should have the sign bit at `1`. So this is equivalent to:
```scala
if (start < 0 || end < 0  || (end - start) < 0 || (cs.length - end) < 0) throw new IndexOutOfBoundsException
```
About the redundancy of the checks, the one in line 83 (`size <= 0`) checks if you're trying to parse just a `"-"` which should fail. Right?

Reply:
Right. The `size <= 0` check should result in an exception thrown for `"-"`.

Credit to [java.io.FilterOutputStream](https://github.com/openjdk/jdk/blob/cec6c068b03d890312e50b448fbc26102c635249/src/java.base/share/classes/java/io/FilterOutputStream.java#L133) for the arcane bitwise bounds check. I assume it was written that way for performance, but I haven't measured if the difference is actually noticeable over your equivalent and clearer version. That class says `@since JDK1.0` which was 1996. CPU branch prediction and the JVM implementation have improved since then.

The bounds check isn't really related to the fix, but since I found one bug and didn't have a good understanding of how this is used, I wanted to nail down the assumptions around inputs. I think `parseLong` is used deep in the parsing code (not public API) so the check could probably be removed if needed, although I didn't expect the cost of keeping it to be significant.